src/query/fuzzy_query.rs

```diff
@@ -129,10 +112,15 @@ impl FuzzyTermQuery {
     }
 
     fn specialized_weight(&self) -> crate::Result<AutomatonWeight<DfaWrapper>> {
-        // LEV_BUILDER is a HashMap, whose `get` method returns an Option
-        match LEV_BUILDER.get(&(self.distance, self.transposition_cost_one)) {
-            // Unwrap the option and build the Ok(AutomatonWeight)
-            Some(automaton_builder) => {
+        static BUILDER: Lazy<RwLock<FxHashMap<(u8, bool), LevenshteinAutomatonBuilder>>> =
+            Lazy::new(Default::default);
+
+        loop {
```

Comment:
Do we really want a RwLock here? Although this makes the logic lazy, it adds additional overhead to every read we try to create the scorer, and I believe the standard library RwLocks still have some performance issues plaguing them when being accessed concurrently by several threads even when reading only, and this is potentially a very hot path.

Secondly, this removes the cap on the maximum edit distance per term, which although isn't bad in itself, does open the potential for people to pick a number *too* high, 3 tends to be the realistic maximum you'd use, although it depends on the word length.

Reply:
> Do we really want a RwLock here? Although this makes the logic lazy, it adds additional overhead to every read we try to create the scorer, and I believe the standard library RwLocks still have some performance issues plaguing them when being accessed concurrently by several threads even when reading only, and this is potentially a very hot path.

Yes, I am somewhat unhappy with the `RwLock` too, but some form of synchronization will have to be necessary if laziness is desired.

Alternatives I see are:
* Using `Mutex` which should be faster but would prevent concurrent reads after the relevant builders have been initialized. Alternatively, `Arc<LevenshteinAutomatonBuilder>` could be cached and cloned under the mutex so that usage can proceed without it.
* But if we do go for the additional indirection of `Arc<LevenshteinAutomatonBuilder>`, we might also use `ArcSwap` instead of `RwLock` (already a dependency) and always publish a new (larger) hash table when another `LevenshteinAutomatonBuilder` had to be initialized containing copies of all the previous `Arc<LevenshteinAutomatonBuilder>` instances. (`LevenshteinAutomatonBuilder` itself is not clone, but I guess that could also be changed upstream to avoid the additional `Arc` indirection in this case.)

> Secondly, this removes the cap on the maximum edit distance per term, which although isn't bad in itself, does open the potential for people to pick a number too high, 3 tends to be the realistic maximum you'd use, although it depends on the word length.

The limitation seems artificial IMHO and related to the eager construction of the cached `LevenshteinAutomatonBuilder` instances which is why I removed it after making things lazy.

I have also read about 3 being a realistic limit, but I think this would better be served by a soft limit, e.g. advice given in the documentation of the `new` and `new_prefix` methods.

Reply:
Opened https://github.com/quickwit-oss/levenshtein-automata/pull/13 to enable cloning of the builders.

Reply:
We could have 2 * 3 static variables and no locks.

Reply:
> We could have a hashmap (or arrays) with once_cell as value? 

We could lazily initialize only the entries and eagerly construct the data structure itself, but if we keep the static limit on the distance, then the current approach - lazily constructing all builders at once - seems to most reasonable to me and I would just remove the TODO.

Reply:
Pushed an implementation of that approach, i.e. using `[[OnceCell<LevenshteinAutomatonBuilder>; 2]; 3]`.